pkg/profileparser/profileparser.go

```diff
@@ -56,6 +60,179 @@ func GetPrefixedName(pbName, objName string) string {
+			}
+			updatedVariable, ok := updated.(*cmpv1alpha1.Variable)
+			if !ok {
+				return fmt.Errorf("unexpected type")
+			}
+
+			foundVariable.Annotations = updatedVariable.Annotations
+			foundVariable.VariablePayload = *updatedVariable.VariablePayload.DeepCopy()
+			return pcfg.Client.Update(context.TODO(), foundVariable)
+		})
+		return err
+	})
+
+	if err != nil {
+		return err
+	}
+
+	if err := deleteObsoleteItems(pcfg.Client, "Variable", pb.Namespace, pb.GetImageDigest()); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+type parsedItemIface interface {
+	metav1.Object
+	k8sruntime.Object
+}
+
+func parseAction(parsedItem parsedItemIface, kind string, pb *cmpv1alpha1.ProfileBundle, pcfg *ParserConfig, updateFn func(found, updated interface{}) error) error {
+	// overwrite name
+	itemName := parsedItem.GetName()
+	parsedItem.SetName(GetPrefixedName(pb.Name, itemName))
+
+	labels := parsedItem.GetLabels()
+	if labels == nil {
+		labels = make(map[string]string)
+	}
+	labels[cmpv1alpha1.ProfileBundleOwnerLabel] = pb.Name
+	parsedItem.SetLabels(labels)
+
+	if err := controllerutil.SetControllerReference(pb, parsedItem, pcfg.Scheme); err != nil {
+		return err
+	}
+
+	key := types.NamespacedName{Name: parsedItem.GetName(), Namespace: parsedItem.GetNamespace()}
+	if err := createOrUpdate(pcfg.Client, kind, key, parsedItem, updateFn); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func createOrUpdate(cli runtimeclient.Client, kind string, key types.NamespacedName, obj k8sruntime.Object, updateFn func(found, updated interface{}) error) error {
+	log.Info("Creating object", "kind", kind, "key", key)
+	found := obj.DeepCopyObject()
+	err := cli.Get(context.TODO(), key, found)
+	if errors.IsNotFound(err) {
+		err := cli.Create(context.TODO(), obj)
+		if err != nil {
+			log.Error(err, "Failed to create object", "kind", kind, "key", key)
+			return err
+		}
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	// Object exist, call up to update
+	if err := updateFn(found, obj); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func deleteObsoleteItems(cli runtimeclient.Client, kind string, namespace string, imageDigest string) error {
+	list := unstructured.UnstructuredList{}
+	list.SetGroupVersionKind(schema.GroupVersionKind{
+		Group:   cmpv1alpha1.SchemeGroupVersion.Group,
+		Version: cmpv1alpha1.SchemeGroupVersion.Version,
+		Kind:    kind + "List",
+	})
+	inNs := runtimeclient.InNamespace(namespace)
+
+	log.Info("Checking for unused object", "kind", kind, "namespace", namespace)
+	if err := cli.List(context.TODO(), &list, inNs); err != nil {
+		return err
+	}
+
+	for _, item := range list.Items {
+		err := deleteIfNotCurrentDigest(cli, imageDigest, &item)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func deleteIfNotCurrentDigest(client runtimeclient.Client, imageDigest string, item *unstructured.Unstructured) error {
+	itemDigest := item.GetAnnotations()[cmpv1alpha1.ProfileImageDigestAnnotation]
```

Comment:
I put the digest into annotations because it can be very long, longer than the limit for labels. This means that we have to iterate over all objects of a type and read their annotations one by one. I thought this was OK, because profile update is a relatively rare operation. The alternative might be to put a shortened version of the digest into the label. e.g. git or quay.io use just first N characters of a digest, but at the same time I think they must have some heuristics to make sure the shortened version is unique. In the end, this is safer, but slower...

Reply:
Agree that this is not frequent enough to worry about optimizing. Might be good to leave a comment about it.

Reply:
Done.